File: src/0261_graph_valid_tree.js
function validTree(n, edges) {
    if (!n) return true;
    const adj = {};
    for (let i = 0; i < n; i++) adj[i] = [];
    for (const [n1, n2] of edges) {
        adj[n1].push(n2);
        adj[n2].push(n1);
    }
    const visit = new Set();
    const dfs = (i, prev) => {
        if (visit.has(i)) return false;
        visit.add(i);
        for (const j of adj[i]) {
            if (j === prev) continue;
            if (!dfs(j, i)) return false;
        }
        return true;
    };
    return dfs(0, -1) && n === visit.size;
}
function validTreeDsu(n, edges) {
    if (n === 0) return true;
    const parents = Array.from({ length: n }, (_, i) => i);
    const ranks = Array(n).fill(1);
    let components = n;
    const find = (node) => {
        while (node !== parents[node]) {
            parents[node] = parents[parents[node]];
            node = parents[node];
        }
        return node;
    };
    const union = (a, b) => {
        const rootA = find(a);
        const rootB = find(b);
        if (rootA === rootB) return false;
        if (ranks[rootA] < ranks[rootB]) parents[rootA] = rootB;
        else if (ranks[rootA] > ranks[rootB]) parents[rootB] = rootA;
        else { parents[rootB] = rootA; ranks[rootA]++; }
        components--;
        return true;
    };
    for (const [a, b] of edges) {
        if (!union(a, b)) return false;
    }
    return components === 1;
}

module.exports = { validTree, validTreeDsu };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(validTree(5, [[0, 1], [0, 2], [0, 3], [1, 4]]), true);
    assert.deepStrictEqual(validTree(5, [[0, 1], [1, 2], [2, 3], [1, 3], [1, 4]]), false);
    assert.deepStrictEqual(validTreeDsu(5, [[0, 1], [0, 2], [0, 3], [1, 4]]), true);
    assert.deepStrictEqual(validTreeDsu(5, [[0, 1], [1, 2], [2, 3], [1, 3], [1, 4]]), false);
}

File: src/0261_graph_valid_tree.rs
struct Solution;

use std::collections::HashSet;

impl Solution {
    fn valid_tree_dfs(i: i32, prev: i32, adj: &[Vec<i32>], visit: &mut HashSet<i32>) -> bool {
        if visit.contains(&i) {
            return false;
        }

        visit.insert(i);
        for &j in &adj[i as usize] {
            if j == prev {
                continue;
            }
            if !Self::valid_tree_dfs(j, i, adj, visit) {
                return false;
            }
        }
        true
    }

    pub fn valid_tree(n: i32, edges: Vec<Vec<i32>>) -> bool {
        if n == 0 {
            return true;
        }

        let mut adj = vec![Vec::new(); n as usize];
        for edge in edges {
            let n1 = edge[0];
            let n2 = edge[1];
            adj[n1 as usize].push(n2);
            adj[n2 as usize].push(n1);
        }

        let mut visit = HashSet::new();
        Self::valid_tree_dfs(0, -1, &adj, &mut visit) && n as usize == visit.len()
    }

    pub fn valid_tree_dsu(n: i32, edges: Vec<Vec<i32>>) -> bool {
        let mut parents: Vec<usize> = (0..n as usize).collect();
        let mut ranks = vec![1; n as usize];
        let mut components = n;

        fn find(value: usize, parents: &mut [usize]) -> usize {
            if parents[value] != value {
                parents[value] = find(parents[value], parents);
            }
            parents[value]
        }

        for edge in edges {
            let root_a = find(edge[0] as usize, &mut parents);
            let root_b = find(edge[1] as usize, &mut parents);
            if root_a == root_b {
                return false;
            }
            if ranks[root_a] > ranks[root_b] {
                parents[root_b] = root_a;
                ranks[root_a] += ranks[root_b];
            } else {
                parents[root_a] = root_b;
                ranks[root_b] += ranks[root_a];
            }
            components -= 1;
        }

        components == 1
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(
            Solution::valid_tree_dsu(5, vec![vec![0, 1], vec![0, 2], vec![0, 3], vec![1, 4]]),
            true
        );
        assert_eq!(
            Solution::valid_tree_dsu(
                5,
                vec![vec![0, 1], vec![1, 2], vec![2, 3], vec![1, 3], vec![1, 4]]
            ),
            false
        );
        assert_eq!(
            Solution::valid_tree(5, vec![vec![0, 1], vec![0, 2], vec![0, 3], vec![1, 4]]),
            true
        );
        assert_eq!(
            Solution::valid_tree(
                5,
                vec![vec![0, 1], vec![1, 2], vec![2, 3], vec![1, 3], vec![1, 4]]
            ),
            false
        );
    }
}
